src/mono/browser/runtime/diagnostics/diag-js.ts

```diff
@@ -0,0 +1,148 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+import { advert1, CommandSetId, dotnet_IPC_V1, ServerCommandId } from "./client-commands";
+import { DiagConnectionBase, downloadBlob, fnClientProvider, IDiagClient, IDiagConnection, IDiagSession, schedule_diagnostic_server_loop, SessionId } from "./common";
+import { PromiseAndController } from "../types/internal";
+import { loaderHelpers } from "./globals";
+import { mono_log_warn } from "./logging";
+import { collectCpuSamples } from "./dotnet-cpu-profiler";
+import { collectPerfCounters } from "./dotnet-counters";
+import { collectGcDump } from "./dotnet-gcdump";
+
+//let diagClient:IDiagClient|undefined = undefined as any;
+//let server:DiagServer = undefined as any;
+
+// configure your application
+// .withEnvironmentVariable("DOTNET_DiagnosticPorts", "download:gcdump")
+// or implement function globalThis.dotnetDiagnosticClient with IDiagClient interface
+
+let nextJsClient:PromiseAndController<IDiagClient>;
+let fromScenarioNameOnce = false;
+
+// Only the last which sent advert is receiving commands for all sessions
+export let serverSession:DiagSession|undefined = undefined;
+
+// singleton wrapping the protocol with the diagnostic server in the Mono VM
+// there could be multiple connection at the same time.
+// DS:advert         ->1
+//                     1<- DC1: command to start tracing session
+// DS:OK, session ID ->1
+// DS:advert         ->2
+// DS:events         ->1
+// DS:events         ->1
+// DS:events         ->1
+// DS:events         ->1
+//                     2<- DC1: command to stop tracing session
+// DS:close          ->1
+
+class DiagSession extends DiagConnectionBase implements IDiagConnection, IDiagSession {
+    public session_id: SessionId = undefined as any;
+    public diagClient?: IDiagClient;
+    public stopDelayedAfterLastMessage:number|undefined = undefined;
+    public resumedRuntime = false;
+
+    constructor (public client_socket:number) {
+        super(client_socket);
+    }
+
+    sendCommand (message: Uint8Array): void {
+        if (!serverSession) {
+            mono_log_warn("no server yet");
+            return;
+        }
+        serverSession.respond(message);
+    }
+
+    async connect_new_client () {
```

Comment:
nit: Naming. So other functions are camelCase. IIRC this is our long term gola

```suggestion
    async connectNewClient () {
```